cloudinit/config/cc_apt_configure.py

```diff
@@ -513,7 +526,6 @@ def add_apt_key(ent, cloud, target=None, hardened=False, file_name=None):
     Supports raw keys or keyid's
     The latter will as a first step fetched to get the raw key
     """
-    _ensure_gpg(cloud)
```

Comment:
This was needed here because `add_mirror_keys` also calls `add_apt_key`. I think we can have it both places if we bring back the LRU cache

Reply:
@TheRealFalcon shoot I missed the add_mirror_keys call as being outside the scope of `add_apt_sources` the problem I'm trying to solve in this PR is that I don't want us to call _ensure_dependencies twice because distro.install_packages will call the expensive `apt update` prior to apt install gpg or software-properties-common. So I pulled that logic up out of `add_apt_key` and into `add_apt_sources`  because I want to do this just once given a condition check on the user-data provided. It looks like I might want to pull the callsite (and user-data inspection of cfg up into `apply_apt` before `add_mirror_keys`. Then, If both gpg requirements and sw-properties-common are going to be required by the provided cloud-config, let's install them in one call to `distro.install_packages`.

Reply:
Not sure this matters to the implementation anymore, but `apt update` is protected by a runner containing a semaphore, which (IIRC) means it will only run once per boot.